### DISCUSSION TITLE
print() doesn't work with custom GFX class

### DISCUSSION THREAD

**Discussion by Spongman** [moononournation/Arduino_GFX#437] [General]
v1.4.4 on rp2040, arduino, platformio.

```
#include <Arduino_GFX_Library.h>

void setup()
{
	class MyGFX : public Arduino_GFX
	{
		int foo;

	public:
		MyGFX(int16_t w, int16_t h) : Arduino_GFX(w, h)
		{
		}

		virtual bool begin(int32_t speed = GFX_NOT_DEFINED) { return true; }

		virtual void writePixelPreclipped(int16_t x, int16_t y, uint16_t color)
		{
			Serial.printf("writePixelPreclipped %i, %i, %x\n", x, y, color);
		}
	};

	MyGFX buffer(240, 135);

	Serial.println(F("start"));
	buffer.startWrite();

	buffer.drawPixel(0, 0, RGB565(255, 255, 0));

	buffer.setCursor(10, 10);
	buffer.setTextColor(RGB565_GREEN);
	buffer.print("text");

	buffer.endWrite();
	Serial.println(F("done"));
}
```

for some reason, `buffer.print()` isn't doing anything, and I can't tell why. all i get is
```
start
writePixelPreclipped 0, 0, ffe0
done
```

If I comment out the `int foo;` line, then it works:
```
start
writePixelPreclipped 0, 0, ffe0
writePixelPreclipped 10, 12, 7e0
writePixelPreclipped 11, 12, 7e0
writePixelPreclipped 12, 10, 7e0
writePixelPreclipped 12, 11, 7e0
writePixelPreclipped 12, 12, 7e0
...
```

**Comment by moononournation** [comment]
you did not mention your method should override parent class

**Comment by Spongman** [comment]
not sure what you mean. the `writePixelPreclipped()` method above _does_ override the one in `Arduino_GFX`

**Comment by moononournation** [comment]
you may refer other class how can override it

**Comment by Spongman** [comment]
> you may refer other class how can override it


sorry, i don't understand what you're saying.


```
#include <iostream>

struct Base
{
	virtual void Method() { std::cout << "base\n"; }
};

struct Derived : public Base
{
	virtual void Method() { std::cout << "derived\n"; }
};

int main()
{
	Base *base = new Derived();
	base->Method();
}
```

prints "derived"

https://godbolt.org/z/M3dKhM397

**Comment by Spongman** [comment]
can you suggest a fix?

**Comment by moononournation** [comment]
no idea

**Comment by Spongman** [comment]
ok, so this 

> you did not mention your method should override parent class

isn't the cause ?

**Reply by moononournation** [reply]
You can see all the child classes have override keyword in header files

**Reply by Spongman** [reply]
the `override` keyword is optional. I clearly illustrated above that `Derived::method()` overrides `Base::Method()`. can you explain adding the `override` keyword does ?

this _issue_ has nothing to do with the `override` keyword and its presence or absence.

**Comment by Spongman** [comment]
why is this not an issue now?

**Reply by moononournation** [reply]
This is not an issue

**Reply by Spongman** [reply]
can you explain why not?

**Reply by Spongman** [reply]
is it _expected_ behavior that the `print()` method does not print anything?

**Reply by moononournation** [reply]
I cannot teach you too much on how to program C++, this just a GFX library.

**Reply by Spongman** [reply]
thanks. i know how to program c++.

**Reply by Spongman** [reply]
can you show how adding/removing the `override` keyword changes the behavior of _either_ of the examples I have given?

**Reply by Spongman** [reply]
this:
```
#include <iostream>

struct Base
{
	virtual void Method() { std::cout << "base\n"; }
};

struct Derived : public Base
{
	void Method() override { std::cout << "derived\n"; }
};

int main()
{
	Base *pb = new Derived();
	pb->Method();
}```

does _exactly_ the same as the example without `override`.

**Reply by moononournation** [reply]
Sorry, I did not have related teaching materials.

**Reply by Spongman** [reply]
there's no teaching required here. however, _you_ seem to be implying that the `override` keyword is necessary when overriding a method in c++. this is not the case. the `override` keyword is 100% optional.

**Reply by Spongman** [reply]
but since you seem to be _stuck_ on that misconception, here's the original example with the `override` keyword added:

```
#include <Arduino_GFX_Library.h>

void setup()
{
	class MyGFX : public Arduino_GFX
	{
		int foo;

	public:
		MyGFX(int16_t w, int16_t h) : Arduino_GFX(w, h)
		{
		}

		bool begin(int32_t speed = GFX_NOT_DEFINED) override { return true; }

		void writePixelPreclipped(int16_t x, int16_t y, uint16_t color) override
		{
			Serial.printf("writePixelPreclipped %i, %i, %x\n", x, y, color);
		}
	};

	MyGFX buffer(240, 135);

	Serial.println(F("start"));
	buffer.startWrite();

	buffer.drawPixel(0, 0, RGB565(255, 255, 0));

	buffer.setCursor(10, 10);
	buffer.setTextColor(RGB565_GREEN);
	buffer.print("text");

	buffer.endWrite();
	Serial.println(F("done"));
}
```

it exhibits exactly the same behavior as I described.

please can you explain how _that_ is not an issue?

**Comment by Spongman** [comment]
this is from chatgpt:

> User: is the override keyword required when overriding a virtual method in c++ ?

> ChatGPT
> No, the override keyword is not strictly required when overriding a virtual method in C++. However, it is considered good practice to use it.
>
> Using the override keyword explicitly tells the compiler that you intend to override a virtual function from a base class. It helps catch errors at compile time, such as misspellings in the function signature or accidental creation of a new function instead of overriding an existing one.
>
> Without the override keyword, the compiler **_will still allow the function to override the virtual function_** if the signatures match, but it won't provide any indication that the function is intended to override a base class function. Therefore, using override improves code readability and helps prevent mistakes.

**Reply by Spongman** [reply]
would you like me to provide more teaching materials?
